automata/fa/fa.py

```diff
@@ -12,3 +20,217 @@ class FA(Automaton, metaclass=abc.ABCMeta):
     """An abstract base class for finite automata."""
 
     __slots__ = tuple()
+
+    @staticmethod
+    def get_state_name(state_data: FAStateT) -> str:
+        """
+        Get an string representation of a state. This is used for displaying and
+        uses `str` for any unsupported python data types.
+        """
+        if isinstance(state_data, str):
+            if state_data == "":
+                return "λ"
+
+            return state_data
+
+        if isinstance(state_data, (set, frozenset, list, tuple)):
+            inner = ", ".join(FA.get_state_name(sub_data) for sub_data in state_data)
+            if isinstance(state_data, (set, frozenset)):
+                if state_data:
+                    return "{" + inner + "}"
+                else:
+                    return "∅"
+
+            if isinstance(state_data, tuple):
+                return "(" + inner + ")"
+
+            if isinstance(state_data, list):
+                return "[" + inner + "]"
+
+        return str(state_data)
+
+    def get_edge_name(self, symbol) -> str:
+        return str(symbol)
+
+    @abc.abstractmethod
+    def iter_states(self) -> Iterable[FAStateT]:
+        """Iterate over all states in the automaton."""
+
+    @abc.abstractmethod
+    def iter_transitions(self) -> Iterable[tuple[FAStateT, FAStateT, Any]]:
```

Comment:
@eliotwrobson ...and I think I can see why, namely because the transition maps for the various automaton types are all slightly different. DFAs don't use sets, but NFAs do, GNFAs are just plain weird lol, and I haven't even gotten to PDAs and TMs yet. lol

So yep, I agree that `iter_transitions` is fine to keep! Resolving this thread.